### ISSUE TITLE
[kicker]: submenus not correctly switched/loaded

### ISSUE THREAD

**Issue by star-buck** [blue-systems/netrunner-15-issues#13]
http://forums.netrunner.com/showthread.php?tid=17376

**Comment by eikehein** [comment]
I can't immediately think of an in-Kicker reason for why the icons would update but the text would not, because the model emits dataChanged() for all data roles at once and the delegate uses the values directly from the model - icons and text aren't treated differently or updated separately, and there's no complicated code involved.

I'll be on the lookout for it whether I can reproduce it (so far I cannot) to investigate further. I thought it was interesting that both menu categories on the shot had the same number of entries and checked whether there might be a funny bug when that happens, but creating some even-numbered submenus to try with didn't help so far ...

**Comment by star-buck** [comment]
I still spot some "overlapping", especially in LIVE mode when quickly going over the main categories with many entries the first time, as if the previous window still has not finished loading, but the next one is already opening. This was much more obvious with the old kde4-based one, but still visible with plasma5 even on quite powerful machines.

**Comment by eikehein** [comment]
This is actually not what's going on though, because there aren't actually two windows -- one major difference between the KDE 4 version of Kicker and the Plasma 5 implementation is that it will reuse (reposition and resize) an existing dialog when switching categories (unless the dialog has to be closed inbetween, e.g. due to an empty category). So what it will do when switching from category A to category B is, depending on whether there are more or fewer items in B than in A, add or chop off items, and update the existing ones, reusing existing delegate instances to avoid delegate instanciation costs. It's all in the service of making category switches as fast and cheap as possible.

The remaining costs are in Plasma's dialog implementation - resizing one is quite costly, although Dialog has been improved massively over 4, which is what made dialog reuse possible at all, in KDE 4 it was too slow for anything - and in KWin (dialog backgrounds are blurred in many themes and the blur shader is complex and expensive).

Anyway, this is why this bug report is a bit mystifying - as per the above explanation, what's going on in the code there is that the existing item delegates are given new data to display, and they are actually handed new icons and new text at the same time (they're data roles on the same model index, which dataChanged gets emitted for), so it's not immediately apparent why one would update but not the other. Bottom line: Needs reproducing to debug since it's nothing obvious (and since it could be outside Kicker, too).

**Comment by llelectronics** [comment]
Hmm... just a short idea as I already had such a problem with Qt5 and QML (maybe it's a bug or race condition). Did you try destroying the model (I guess you are using one) and recreate it. I know maybe expensive aswell but could fix the bug maybe.

**Comment by eikehein** [comment]
No, and that would be wrong to do anyway since it does away with the intended benefits :).

The Kicker model backend works like this, for the record: The menu hierarchy is a list model of list models, and the dialog is directly using a custom funnel proxy model that has its source model swapped out when the category is changed. The funnel takes care of figuring out the difference between the two lists and emitting dataChanged and rowsInserted/rowsRemoved as necessary.

Taking out the funnel and going to "swap out the source model of the dialog directly" is two-lines of code change, but if were to help with this problem, it would be because the model swap causes all delegates to be destroyed and reinstanciated, which is exactly what we want to avoid: The funnel is there to make sure that existing delegates are reused and just get new data (in response to dataChanged) for the least amount of work and maximum performance.

If this bug were in Kicker, the obvious place where the bug would be is that funnel: It could be calculating wrong dataChanged emits. However, the screenshots show that the icons do update, and the funnel doesn't specify any data roles when it emits, so the view should update the data for all roles. That's why I currently have no idea until/if I can ever reproduce this.
